tests/framework/scheduler.py

```diff
@@ -0,0 +1,329 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Pytest plugin that schedules tests to run concurrently.
+
+This plugin adds a new command line option (`--concurrency`), allowing the
+user to choose the maximum number of worker processes that can run tests
+concurrently.
+
+Tests are split into batches, each batch being assigned a maximum concurrency
+level. For instance, all performance tests will run sequentially
+(i.e. concurrency=1), since they rely on the availability of the full host
+resources, in order to make accurate measurements. Additionally, other tests
+may be restricted to running sequentially, if they are per se
+concurrency-unsafe. See `PytestScheduler.pytest_runtestloop()`.
+
+Scheduling is achieved by overriding the pytest run loop (i.e.
+`pytest_runtestloop()`), and splitting the test session item list across
+multiple `fork()`ed worker processes. Since no user code is run before
+`pytest_runtestloop()`, each worker becomes a pytest session itself.
+Reporting is disabled for worker process, each worker sending its results
+back to the main / server process, via an IPC pipe, for aggregation.
+"""
+
+import multiprocessing as mp
+import os
+import re
+import sys
+from random import random
+from select import select
+from time import sleep
+
+from . import mpsing  # pylint: disable=relative-beyond-top-level
+
+
+class PytestScheduler(mpsing.MultiprocessSingleton):
+    """A pretty custom test execution scheduler."""
+
+    def __init__(self):
+        """Initialize the scheduler.
+
+        Not to be called directly, since this is a singleton. Use
+        `PytestScheduler.instance()` to get the scheduler object.
+        """
+        super().__init__()
+        self._mpsings = [self]
+        self.session = None
+
+    def register_mpsing(self, mpsing_inst):
+        """Register a multi-process singleton object.
+
+        Since the scheduler will be handling the main testing loop, it needs
+        to be aware of any multi-process singletons that must be serviced
+        during the test run (i.e. polled and allowed to handle method
+        execution in the server context).
+        """
+        self._mpsings.append(mpsing_inst)
+
+    @staticmethod
+    def do_pytest_addoption(parser):
+        """Pytest hook. Add concurrency command line option."""
+        avail_cpus = len(os.sched_getaffinity(0))
+        # Defaulting to a third of the available (logical) CPUs sounds like a
+        # good enough plan.
+        default = max(1, int(avail_cpus / 3))
+        parser.addoption(
+            "--concurrency",
```

Comment:
I don't think you need to write the option twice. Do you?

Reply:
As far as I understand, there are two args there to allow long-form and short-form aliasing (i.e. `-X` and `--longer-description-of-X`).